storage.go

```diff
@@ -2,35 +2,233 @@ package gdrive
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	o = s.newObject(true)
+	o.ID = path
+	o.Path = path
+	o.Mode = ModeDir
+
+	return o, nil
+
 }
 
 func (s *Storage) delete(ctx context.Context, path string, opt pairStorageDelete) (err error) {
-	panic("not implemented")
+	var fileId string
+	fileId, err = s.pathToId(ctx, path)
+	if err != nil {
+		return err
+	}
+	err = s.service.Files.Delete(fileId).Do()
+	if err != nil {
+		return err
+	}
+	return nil
 }
 
 func (s *Storage) list(ctx context.Context, path string, opt pairStorageList) (oi *ObjectIterator, err error) {
-	panic("not implemented")
+	input := &objectPageStatus{
+		limit: 200,
+		path:  s.getAbsPath(path),
+	}
+
+	if !opt.HasListMode || opt.ListMode.IsDir() {
+		return NewObjectIterator(ctx, s.nextObjectPage, input), nil
+	} else {
+		return nil, services.ListModeInvalidError{Actual: opt.ListMode}
+	}
 }
 
 func (s *Storage) metadata(opt pairStorageMetadata) (meta *StorageMeta) {
-	panic("not implemented")
+	meta = NewStorageMeta()
+	meta.Name = s.name
+	meta.WorkDir = s.workDir
+	return meta
+}
+
+// Create a directory by passing it's name and the parents' fileId.
+// It will return the fileId of the directory whether it exist or not.
+// If error occurs, it will return an empty string and error.
+func (s *Storage) mkDir(ctx context.Context, parents string, dirName string) (string, error) {
+	id, err := s.searchContentInDir(ctx, parents, dirName)
+	if err != nil {
+		return "", err
+	}
+	// Simply return the fileId if the directory already exist
+	if id != "" {
+		return id, nil
+	}
+
+	// create a directory if not exist
+	dir := &drive.File{
+		Name:     dirName,
+		Parents:  []string{parents},
+		MimeType: DirectoryMimeType,
+	}
+	f, err := s.service.Files.Create(dir).Context(ctx).Do()
+	if err != nil {
+		return "", err
+	}
+	return f.Id, nil
+}
+
+func (s *Storage) nextObjectPage(ctx context.Context, page *ObjectPage) error {
+	input := page.Status.(*objectPageStatus)
+	dirId, err := s.pathToId(ctx, input.path)
+	if err != nil {
+		return err
+	}
+	q := s.service.Files.List().Q(fmt.Sprintf("parents=%s", dirId))
+
+	if input.pageToken != "" {
+		q = q.PageToken(input.pageToken)
+	}
+	r, err := q.Do()
+
+	if err != nil {
+		return err
+	}
+	for _, f := range r.Files {
+		o := s.newObject(true)
+		// There is no way to get the path of the file directly, we have to do this
+		o.Path = input.path + "/" + f.Name
+		switch f.MimeType {
+		case DirectoryMimeType:
+			o.Mode = ModeDir
+		default:
+			o.Mode = ModeRead
+		}
+		page.Data = append(page.Data, o)
+	}
+
+	input.pageToken = r.NextPageToken
+	return nil
+}
+
+// TODO: add cache support
+// pathToId converts path to fileId, as we talked in RFC-14.
+// Ref: https://github.com/beyondstorage/go-service-gdrive/blob/master/docs/rfcs/14-gdrive-for-go-storage-design.md
+// If success, we will return the fileId of the path we passing, and nil for sure.
+// But if the path is not exist, we will return an empty string and error.
```

Comment:
we will send API in `pathToId`, how will we handle the API error here?

How about adding a rule that:

- `err` represents the error handled in `pathToId`
- `fileId` represents the results: `fileId` empty means the path is not exist, otherwise it's the `fileId` of input path.